### PR TITLE
9228 - add OIDC development setup for OIDC login feature testing

### LINKED ISSUE #9228: As a developer, I'd like Keycloak or similar so I can test OIDC login
As discussed just now with @poikilotherm @GPortas @johannes-darms and @ekraffmiller developers currently don't have a way to test the [login with OIDC](https://guides.dataverse.org/en/5.12.1/installation/oidc.html) feature of Dataverse.

Definition of done (assuming Keycloak):

- Add KeyCloak config and document setup at (future version of) https://guides.dataverse.org/en/5.12.1/developers/remote-users.html
- (Optional) add some automation of the setup with Docker or similar.

### PR DESCRIPTION
**Pull request by GPortas**
**What this PR does / why we need it**:

As stated in #9228, developers currently don't have a way to test the [login with OIDC](https://guides.dataverse.org/en/5.12.1/installation/oidc.html) feature of Dataverse.

This PR includes a dockerized Keycloak setup with OIDC support for development purposes, as well as its associated documentation.

TODOs.

- [X] Add Keycloak-Docker config
- [x] Add Keycloak-Docker documentation at (future version of) https://guides.dataverse.org/en/5.12.1/developers/remote-users.html
- [x] Test current OIDC login flow with this setup

**Which issue(s) this PR closes**: 

Closes #9228

**Special notes for your reviewer**:
Not yet.

**Suggestions on how to test this**:

Follow the next steps:

1. (Positioned in `/conf/keycloak/`) Run Keycloak docker container: You can use _docker-compose_ file or _run-keycloak.sh_ script. 
2. Verify that Keycloak is running on http://localhost:8090/.
3. Verify that Dataverse is locally running
4. (Positioned in `/conf/keycloak/`) Execute the following API call which enables the Keycloak OIDC client as an authentication provider for Dataverse:  `curl -X POST -H 'Content-type: application/json' --upload-file oidc-keycloak-auth-provider.json http://localhost:8080/api/admin/authenticationProviders`
5. Test that the new OIDC auth flow works. A new option "OIDC-Keycloak" should appear on the login screen. The credentials for the Keycloak test user are: kcuser / kcpassword (usr/pwd).

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:
No.

**Is there a release notes update needed for this change?**:
Not sure.

**Additional documentation**:
Not yet.

### CODE REVIEW

**Comment by pdurbin** [comment]
I gave this a size of 3. As long as a developer has Docker installed, it should be pretty easy to spin up Keycloak, add the config, and to a quick test. The docs @GPortas wrote are excellent.

**Comment by mreekie** [comment]
Prioritization note:
@siacus This is work that is coming out of the re-architecture work.  
We haven't discussed yet how we're going to priortize these so for now I've added it directly to the ordered backlog.

File: conf/keycloak/oidc-keycloak-auth-provider.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "oidc-keycloak",
+  "factoryAlias": "oidc",
+  "title": "OIDC-Keycloak",
+  "subtitle": "OIDC-Keycloak",
+  "factoryData": "type: oidc | issuer: http://localhost:8090/auth/realms/oidc-realm | clientId: oidc-client | clientSecret: ss6gE8mODCDfqesQaSG3gwUwZqZt547E",
```
Comment (qqmyers) [inline conf/keycloak/oidc-keycloak-auth-provider.json:6]: not so secret in github?

Comment (GPortas) [reply conf/keycloak/oidc-keycloak-auth-provider.json:6]: This setup is intended for local testing only. For a remote or production OIDC provider, this configuration file, as well as the Realm file (if using Keycloak for OIDC) should not be included in the repository.

**Comment by mreekie** [comment]
> Prioritization note: @siacus This is work that is coming out of the re-architecture work. We haven't discussed yet how we're going to priortize these so for now I've added it directly to the ordered backlog.

Corrected. This is part of: "NIH OTA: 1.7.2" and so is part of the NIH Backlog items.

File: conf/keycloak/oidc-keycloak-auth-provider.json (hunk 1 above)
Comment (pdurbin) [reply conf/keycloak/oidc-keycloak-auth-provider.json:6]: It's fine. Dev only. Never prod, please. I added a warning about this: bb625ea

In the future, assuming we ship a Keycloak config for prod, can worry about secrets then.

**Review by pdurbin — APPROVED** [review]
I tested this and it works amazingly well. Thank you, @GPortas!! I've wanted something like this for over three years, when OIDC support was added in PR #6433 by @poikilotherm and I never had an easy way to test it myself.

It's inconvenient to add screenshots and details in the review box but I'll add another comment in a minute. I did make a couple tiny doc tweaks.

Approved. Thanks again! 🎉 🚀

**Comment by pdurbin** [comment]
This pull request is a dream come true. Just works!

I did make a couple tiny tweaks to the docs (formatting stuff and a security warning not to run this config it prod since the client secret is in GitHub).

The Bash option was listed first so I tried it first. Worked great. Here's the output:

$ cd conf/keycloak
$ ./run-keycloak.sh
Unable to find image 'jboss/keycloak:16.1.1' locally
16.1.1: Pulling from jboss/keycloak
ac10f00499d5: Pull complete 
96d53117c12e: Pull complete 
1d929376eb7f: Pull complete 
93e1e1b6d192: Pull complete 
f353ba0db29e: Pull complete 
Digest: sha256:abdb1aea6c671f61a594af599f63fbe78c9631767886d9030bc774d908422d0a
Status: Downloaded newer image for jboss/keycloak:16.1.1
cc78453726bb8292e241472b3d30351e651d8a35a431ad5ea4c7f7063787842b
INFO - Keycloak container created and running

I got a working Keycloak instance at http://localhost:8090 which looked like this:

Screen Shot 2023-01-06 at 2 54 34 PM

Then I loaded up the auth provider:

$ curl -X POST -H 'Content-type: application/json' --upload-file oidc-keycloak-auth-provider.json http://localhost:8080/api/admin/authenticationProviders
{"status":"OK","data":{"id":"oidc-keycloak","factoryAlias":"oidc","title":"OIDC-Keycloak","subtitle":"OIDC-Keycloak","factoryData":"type: oidc | issuer: http://localhost:8090/auth/realms/oidc-realm | clientId: oidc-client | clientSecret: ss6gE8mODCDfqesQaSG3gwUwZqZt547E","enabled":true}}

Here are screenshots of the login process:

Screen Shot 2023-01-06 at 2 55 51 PM
Screen Shot 2023-01-06 at 2 56 05 PM
Screen Shot 2023-01-06 at 2 56 34 PM
Screen Shot 2023-01-06 at 2 56 56 PM
Screen Shot 2023-01-06 at 2 57 31 PM
Screen Shot 2023-01-06 at 2 57 47 PM
Screen Shot 2023-01-06 at 2 58 11 PM
Screen Shot 2023-01-06 at 2 58 47 PM

One thing to note above is that no username was prepopulated (given name, family name, and email were). @poikilotherm already wrote this bug up here:

- #6690

This PR means that issue will be MUCH easier for a developer to work on! 🎉 

During the auth meeting today I said I was curious what gets stored as a `persistentUserId` for an Keycloak user. It turns out that it's a UUID, like this: `"persistentUserId": "47411075-985b-4587-bbde-167e3fd8c949"`

Here's the full output of a dump of this keycloak user:

```
$ curl -s http://localhost:8080/api/admin/authenticatedUsers/kcuser | jq .
{
  "status": "OK",
  "data": {
    "id": 2,
    "identifier": "@kcuser",
    "displayName": "Test Test",
    "firstName": "Test",
    "lastName": "Test",
    "email": "test@test.com",
    "superuser": false,
    "deactivated": false,
    "persistentUserId": "47411075-985b-4587-bbde-167e3fd8c949",
    "createdTime": "2023-01-06T19:57:37Z",
    "lastLoginTime": "2023-01-06T19:57:37Z",
    "authenticationProviderId": "oidc-keycloak"
  }
}
```

I also tested the rm-keycloak.sh script and the docker compose option. It all worked great. Thanks again, @GPortas!

**Comment by GPortas** [comment]
Very nice step-by-step guide with screenshots and good information about `persistentUserId`.

Thank you @pdurbin!

**Comment by mreekie** [comment]
Updated information.
- This was an add-on from Phil for the "free for all" sprint

**Comment by pdurbin** [comment]
I said I would pick this up during the auth meeting last week as we discussed the [Dataverse - SPA Authentication V1](https://docs.google.com/document/d/1mzPLbzAiED93hTXnTyw9rDBm7_dL1Qf4bPMg2j0c0NE/edit?usp=sharing) doc.

It seemed absolutely ready to merge. High value (extremely useful for ongoing auth work). Zero risk. So I went ahead and moved it to QA. Next time, I'm happy to simply merge it if that's better. I was certainly tempted! 😄

**Pull request merged**